<a href="https://colab.research.google.com/github/Aleeha-Fatima-del/task01_aleeha-fatima/blob/main/Week__1_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

### Load the dataset

In [ ]:
file_path = '/content/Dataset for Data Analytics (1).xlsx'
df = pd.read_excel(file_path)

print("Dataset loaded successfully. Displaying the first 5 rows:")
display(df.head())

Dataset loaded successfully. Displaying the first 5 rows:


,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10
1,ORD200001,2024-08-23,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,SAVE10,Referral,302.70
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40
3,ORD200003,2023-10-15,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,SAVE10,Facebook,273.19
4,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04


### Identify Missing Values

In [ ]:
missing_values = df.isnull().sum()
missing_values = missing_values[missing_values > 0]

if missing_values.empty:
    print("No missing values found in the dataset.")
else:
    print("Columns with missing values and their counts:")
    display(missing_values)

    print("\nSuggesting imputation strategies:")
    for column, count in missing_values.items():
        if df[column].dtype == 'object':
            print(f"- Column '{column}' (categorical): Impute with mode.")
        elif df[column].dtype in ['int64', 'float64']:
            print(f"- Column '{column}' (numerical): Impute with mean or median, depending on distribution (check for skewness). Mean is generally preferred for symmetric data, median for skewed data.")
        elif df[column].dtype == 'datetime64[ns]':
            print(f"- Column '{column}' (date/time): Impute with a specific date (e.g., mode, or context-dependent value) if appropriate, or consider dropping rows if the missing dates are critical.")

Columns with missing values and their counts:


,0
CouponCode,309



Suggesting imputation strategies:
- Column 'CouponCode' (categorical): Impute with mode.


### Impute Missing Values

In [ ]:
mode_coupon_code = df['CouponCode'].mode()[0]
df['CouponCode'] = df['CouponCode'].fillna(mode_coupon_code)

print(f"Missing values in 'CouponCode' imputed with the mode: {mode_coupon_code}")

print("\nDataFrame after imputation (first 5 rows):")
display(df.head())

print("\nRe-checking for missing values:")
missing_values_after_imputation = df.isnull().sum()
missing_values_after_imputation = missing_values_after_imputation[missing_values_after_imputation > 0]

if missing_values_after_imputation.empty:
    print("No missing values found in the dataset after imputation.")
else:
    print("Remaining columns with missing values and their counts after imputation:")
    display(missing_values_after_imputation)

Missing values in 'CouponCode' imputed with the mode: FREESHIP

DataFrame after imputation (first 5 rows):


,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10
1,ORD200001,2024-08-23,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,SAVE10,Referral,302.70
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40
3,ORD200003,2023-10-15,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,SAVE10,Facebook,273.19
4,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04



Re-checking for missing values:
No missing values found in the dataset after imputation.


### Remove Duplicates

In [ ]:
initial_rows = df.shape[0]
print(f"Initial number of rows: {initial_rows}")

duplicate_order_ids = df[df.duplicated(subset='OrderID', keep=False)]
if not duplicate_order_ids.empty:
    print("\nDuplicate OrderIDs found. Displaying duplicate records:")
    display(duplicate_order_ids.sort_values(by='OrderID'))

    df.drop_duplicates(subset='OrderID', keep='first', inplace=True)
    rows_after_dropping = df.shape[0]
    duplicates_removed = initial_rows - rows_after_dropping
    print(f"\n{duplicates_removed} duplicate rows removed based on 'OrderID'.")
    print(f"Number of rows after removing duplicates: {rows_after_dropping}")
else:
    print("\nNo duplicate OrderIDs found.")

print("\nDataFrame after duplicate removal (first 5 rows):")
display(df.head())

Initial number of rows: 1200

No duplicate OrderIDs found.

DataFrame after duplicate removal (first 5 rows):


,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10
1,ORD200001,2024-08-23,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,SAVE10,Referral,302.70
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40
3,ORD200003,2023-10-15,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,SAVE10,Facebook,273.19
4,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04


### Correct Data Formats

In [ ]:
# Convert 'Date' column to ISO 8601 format (YYYY-MM-DD)
# Errors='coerce' will turn unparseable dates into NaT (Not a Time)
df['Date'] = pd.to_datetime(df['Date'], errors='coerce').dt.strftime('%Y-%m-%d')

print("Date column converted to YYYY-MM-DD format (first 5 rows):")
display(df[['Date']].head())

Date column converted to YYYY-MM-DD format (first 5 rows):


,Date
0,2023-01-04
1,2024-08-23
2,2024-02-27
3,2023-10-15
4,2025-05-08


In [ ]:
# Ensure numeric columns (Quantity, UnitPrice, TotalPrice) have consistent precision (2 decimals)
numeric_cols = ['Quantity', 'UnitPrice', 'TotalPrice']
for col in numeric_cols:
    if col in df.columns:
        # Convert to numeric first, coercing errors to NaN
        df[col] = pd.to_numeric(df[col], errors='coerce')
        # Fill potential NaNs from coercion if necessary (e.g., with 0 or mean/median)
        # For this task, we assume valid numbers or will keep NaN if conversion fails.
        # Format to 2 decimal places. Using round() for numerical operations.
        # This keeps the column as a numeric type for further calculations.
        df[col] = df[col].round(2)

print("\nNumeric columns formatted to 2 decimal places (first 5 rows of relevant columns):")
display(df[numeric_cols].head())


Numeric columns formatted to 2 decimal places (first 5 rows of relevant columns):


,Quantity,UnitPrice,TotalPrice
0,5,570.62,2853.10
1,2,151.35,302.70
2,5,550.68,2753.40
3,1,273.19,273.19
4,4,626.01,2504.04


In [ ]:
# Trim whitespace and apply proper case to text fields
text_cols = ['Product', 'ShippingAddress', 'PaymentMethod', 'ReferralSource', 'OrderStatus']
for col in text_cols:
    if col in df.columns:
        # Convert to string type first to ensure .str accessor is available
        df[col] = df[col].astype(str).str.strip().str.title()

print("\nText fields cleaned (trimmed whitespace and proper cased) (first 5 rows of relevant columns):")
display(df[text_cols].head())


Text fields cleaned (trimmed whitespace and proper cased) (first 5 rows of relevant columns):


,Product,ShippingAddress,PaymentMethod,ReferralSource,OrderStatus
0,Monitor,928 Main St,Debit Card,Instagram,Shipped
1,Phone,823 Main St,Online,Referral,Shipped
2,Tablet,512 Main St,Credit Card,Email,Cancelled
3,Chair,275 Main St,Debit Card,Facebook,Returned
4,Printer,668 Main St,Online,Email,Delivered


### Verify Data Formats after Correction

In [ ]:
print("\nUpdated DataFrame info:")
df.info()

print("\nFirst 5 rows of the DataFrame after all format corrections:")
display(df.head())


Updated DataFrame info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   OrderID          1200 non-null   object 
 1   Date             1200 non-null   object 
 2   CustomerID       1200 non-null   object 
 3   Product          1200 non-null   object 
 4   Quantity         1200 non-null   int64  
 5   UnitPrice        1200 non-null   float64
 6   ShippingAddress  1200 non-null   object 
 7   PaymentMethod    1200 non-null   object 
 8   OrderStatus      1200 non-null   object 
 9   TrackingNumber   1200 non-null   object 
 10  ItemsInCart      1200 non-null   int64  
 11  CouponCode       1200 non-null   object 
 12  ReferralSource   1200 non-null   object 
 13  TotalPrice       1200 non-null   float64
dtypes: float64(2), int64(2), object(10)
memory usage: 131.4+ KB

First 5 rows of the DataFrame after all format corrections:


,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10
1,ORD200001,2024-08-23,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,SAVE10,Referral,302.70
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40
3,ORD200003,2023-10-15,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,SAVE10,Facebook,273.19
4,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04


### 4. Validate Data Ranges

In [ ]:
# Validate numerical columns for reasonable ranges (e.g., non-negative)
# Quantity, UnitPrice, TotalPrice, ItemsInCart should ideally be positive

numerical_range_cols = ['Quantity', 'UnitPrice', 'TotalPrice', 'ItemsInCart']

for col in numerical_range_cols:
    if col in df.columns:
        # Convert to numeric first, coercing errors to NaN
        df[col] = pd.to_numeric(df[col], errors='coerce')

        # Check for negative values
        negative_values = df[df[col] < 0]
        if not negative_values.empty:
            print(f"\nFound negative values in column '{col}':")
            display(negative_values)
            print(f"Suggesting: Investigate these records. Depending on context, they might be errors, or represent returns/discounts. For now, we will mark them as NaN to be addressed.")
            df.loc[df[col] < 0, col] = None # Or np.nan for numerical columns

        # Check for zero or unusually high values (simple outlier check for demonstration)
        # For Quantity and ItemsInCart, zero might be acceptable. For prices, it's usually not.
        if col in ['UnitPrice', 'TotalPrice']:
            zero_values = df[df[col] == 0]
            if not zero_values.empty:
                print(f"\nFound zero values in column '{col}':")
                display(zero_values)
                print(f"Suggesting: Investigate these records. Zero prices usually indicate errors or specific business logic (e.g., free items). For now, we will mark them as NaN.")
                df.loc[df[col] == 0, col] = None

# Validate 'Date' column for future dates
# Convert 'Date' column to datetime objects first for comparison
df['Date'] = pd.to_datetime(df['Date'], errors='coerce')

current_date = pd.to_datetime('today').normalize()
future_dates = df[df['Date'] > current_date]

if not future_dates.empty:
    print(f"\nFound future dates in the 'Date' column:")
    display(future_dates)
    print(f"Suggesting: Investigate these records. Future dates might be data entry errors. For now, we will mark them as NaT (Not a Time).")
    df.loc[df['Date'] > current_date, 'Date'] = pd.NaT
else:
    print("\nNo future dates found in the 'Date' column.")

print("\nDataFrame head after range validation (showing relevant columns):")
display(df[['Date'] + numerical_range_cols].head())


No future dates found in the 'Date' column.

DataFrame head after range validation (showing relevant columns):


,Date,Quantity,UnitPrice,TotalPrice,ItemsInCart
0,2023-01-04,5,570.62,2853.10,7
1,2024-08-23,2,151.35,302.70,3
2,2024-02-27,5,550.68,2753.40,8
3,2023-10-15,1,273.19,273.19,5
4,2025-05-08,4,626.01,2504.04,8


### 5. Audit Integrity

In [ ]:
# Calculate the expected TotalPrice
# Ensure Quantity and UnitPrice are numeric before multiplication
df['ExpectedTotalPrice'] = df['Quantity'] * df['UnitPrice']

# Define a small tolerance for floating-point comparisons
tolerance = 0.01 # 2 decimal places accuracy

# Flag rows where TotalPrice does not match Quantity * UnitPrice within tolerance
discrepancy_rows = df[abs(df['TotalPrice'] - df['ExpectedTotalPrice']) > tolerance]

if not discrepancy_rows.empty:
    print("\nDiscrepancies found where TotalPrice != Quantity * UnitPrice (within tolerance):")
    display(discrepancy_rows[['Quantity', 'UnitPrice', 'TotalPrice', 'ExpectedTotalPrice']])

    # Correct inconsistencies: update TotalPrice to ExpectedTotalPrice
    df.loc[abs(df['TotalPrice'] - df['ExpectedTotalPrice']) > tolerance, 'TotalPrice'] = df['ExpectedTotalPrice']
    print(f"\n{len(discrepancy_rows)} inconsistencies corrected. TotalPrice updated to Quantity * UnitPrice.")
else:
    print("\nNo discrepancies found between TotalPrice and Quantity * UnitPrice.")

# Drop the temporary 'ExpectedTotalPrice' column
df.drop(columns=['ExpectedTotalPrice'], inplace=True)

print("\nDataFrame head after integrity audit (showing relevant columns):")
display(df[['Quantity', 'UnitPrice', 'TotalPrice']].head())


No discrepancies found between TotalPrice and Quantity * UnitPrice.

DataFrame head after integrity audit (showing relevant columns):


,Quantity,UnitPrice,TotalPrice
0,5,570.62,2853.10
1,2,151.35,302.70
2,5,550.68,2753.40
3,1,273.19,273.19
4,4,626.01,2504.04


### 5. Standardize Categories

In [ ]:
# Standardize 'PaymentMethod'
# Convert to title case and then map to desired consistent values
df['PaymentMethod'] = df['PaymentMethod'].str.title()
payment_method_mapping = {
    'Credit Card': 'Credit Card',
    'Debit Card': 'Debit Card',
    'Paypal': 'PayPal',
    'Bank Transfer': 'Bank Transfer'
}
df['PaymentMethod'] = df['PaymentMethod'].replace(payment_method_mapping)

print("Unique values in 'PaymentMethod' after standardization:")
print(df['PaymentMethod'].unique())

# Standardize 'OrderStatus'
df['OrderStatus'] = df['OrderStatus'].str.title()
order_status_mapping = {
    'Delivered': 'Delivered',
    'Shipped': 'Shipped',
    'Cancelled': 'Cancelled',
    'Returned': 'Returned',
    'Pending': 'Pending'
}
df['OrderStatus'] = df['OrderStatus'].replace(order_status_mapping)

print("\nUnique values in 'OrderStatus' after standardization:")
print(df['OrderStatus'].unique())

# Standardize 'ReferralSource'
df['ReferralSource'] = df['ReferralSource'].str.title()
referral_source_mapping = {
    'Instagram': 'Instagram',
    'Facebook': 'Facebook',
    'Google': 'Google',
    'Email': 'Email',
    'Referral': 'Referral'
}
df['ReferralSource'] = df['ReferralSource'].replace(referral_source_mapping)

print("\nUnique values in 'ReferralSource' after standardization:")
print(df['ReferralSource'].unique())

print("\nDataFrame head after categorical standardization (showing relevant columns):")
display(df[['PaymentMethod', 'OrderStatus', 'ReferralSource']].head())

Unique values in 'PaymentMethod' after standardization:
['Debit Card' 'Online' 'Credit Card' 'Gift Card' 'Cash']

Unique values in 'OrderStatus' after standardization:
['Shipped' 'Cancelled' 'Returned' 'Delivered' 'Pending']

Unique values in 'ReferralSource' after standardization:
['Instagram' 'Referral' 'Email' 'Facebook' 'Google']

DataFrame head after categorical standardization (showing relevant columns):


,PaymentMethod,OrderStatus,ReferralSource
0,Debit Card,Shipped,Instagram
1,Online,Shipped,Referral
2,Credit Card,Cancelled,Email
3,Debit Card,Returned,Facebook
4,Online,Delivered,Email


### 6. Document Changes

| Change ID | Description                                                                        | Impact                                                                                                                                                                          | Status    |
|-----------|------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------|
| CR001     | Imputed missing `CouponCode` values with the mode.                                 | Ensured all rows have a value for `CouponCode`, preventing potential analysis issues from missing data. 309 missing values were filled with 'FREESHIP'.                          | Resolved  |
| CR002     | Checked for duplicate `OrderID`s.                                                    | Confirmed no duplicate `OrderID`s existed in the dataset, ensuring each order is unique.                                                                                        | Resolved  |
| CR003     | Standardized `Date` column to ISO 8601 format (YYYY-MM-DD).                        | Provided a consistent date format for easier time-series analysis and compatibility.                                                                                            | Resolved  |
| CR004     | Formatted numeric columns (`Quantity`, `UnitPrice`, `TotalPrice`) to two decimal places. | Ensured numerical precision and consistency across financial data points.                                                                                                       | Resolved  |
| CR005     | Cleaned text fields (`Product`, `ShippingAddress`, `PaymentMethod`, `ReferralSource`, `OrderStatus`). | Trimmed whitespace and applied title case for consistent text representation, improving data quality and readability.                                                               | Resolved  |
| CR006     | Validated data ranges for numerical columns (`Quantity`, `UnitPrice`, `TotalPrice`, `ItemsInCart`) and `Date` column. | Confirmed no negative values in numerical columns, no zero prices, and no future dates. This verified the logical integrity of key data points.                                   | Resolved  |
| CR007     | Audited `TotalPrice` against `Quantity × UnitPrice`.                               | Verified the accuracy of the `TotalPrice` column by comparing it to calculated values; no discrepancies were found, confirming data consistency for these financial records. | Resolved  |
| CR008     | Standardized categorical values for `PaymentMethod`, `OrderStatus`, and `ReferralSource`. | Unified categories by converting to title case and mapping specific values, improving consistency for analysis and reducing data entry variations.                              | Resolved  |

### 7. Final Verification

In [ ]:
# 7.1 Verify Unique Identifiers (OrderID)

num_unique_order_ids = df['OrderID'].nunique()
total_order_ids = len(df['OrderID'])

if num_unique_order_ids == total_order_ids:
    print(f"Verification: All {num_unique_order_ids} OrderIDs are unique (0% error rate).")
else:
    print(f"Verification: Found {total_order_ids - num_unique_order_ids} duplicate OrderIDs.")


Verification: All 1200 OrderIDs are unique (0% error rate).


In [ ]:
# 7.2 Verify Date Formats (ISO 8601)

# The 'Date' column was converted to datetime objects during data validation,
# so we need to ensure it's still treated as a string in YYYY-MM-DD for this verification.
# Re-convert to string in the desired format if it's not already.

# First, ensure 'Date' is datetime type for robust format checking
df['Date'] = pd.to_datetime(df['Date'], errors='coerce')

# Check if all non-NaT dates conform to the 'YYYY-MM-DD' string format
# We check for NaT because `pd.to_datetime` might have coerced some invalid dates
iso_8601_dates = df['Date'].dropna().apply(lambda x: x.strftime('%Y-%m-%d')).apply(lambda x: pd.to_datetime(x, errors='coerce').strftime('%Y-%m-%d') == x)

if iso_8601_dates.all():
    print("Verification: All dates in the 'Date' column are in ISO 8601 (YYYY-MM-DD) format (0% error rate).")
else:
    print(f"Verification: Found {len(iso_8601_dates) - iso_8601_dates.sum()} dates not in ISO 8601 format.")




Verification: All dates in the 'Date' column are in ISO 8601 (YYYY-MM-DD) format (0% error rate).


In [ ]:
# Summary of Cleaned Dataset

print(f"\nSummary of Cleaned Dataset:")
print(f"Total rows: {len(df)}")
print("Sample preview of the cleaned dataset (first 5 rows):")
display(df.head())


Summary of Cleaned Dataset:
Total rows: 1200
Sample preview of the cleaned dataset (first 5 rows):


,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10
1,ORD200001,2024-08-23,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,SAVE10,Referral,302.70
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40
3,ORD200003,2023-10-15,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,SAVE10,Facebook,273.19
4,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04
